plugin/httptrace/clienttrace.go

```diff
@@ -76,25 +91,32 @@ func NewClientTrace(ctx context.Context) *httptrace.ClientTrace {
 }
 
 func (ct *clientTracer) start(hook, spanName string, attrs ...core.KeyValue) {
-	_, sp := ct.tr.Start(ct.Context, spanName, trace.WithAttributes(attrs...), trace.WithSpanKind(trace.SpanKindClient))
 	ct.mtx.Lock()
 	defer ct.mtx.Unlock()
-	if ct.root == nil {
-		ct.root = sp
+
+	ctx, ok := ct.activeHooks[parentHook(hook)]
+	if !ok {
+		ctx = ct.Context
 	}
-	if _, ok := ct.activeHooks[hook]; ok {
-		// end was called before start is handled.
-		sp.End()
-		delete(ct.activeHooks, hook)
+
+	if hookCtx, found := ct.activeHooks[hook]; !found {
+		var sp trace.Span
+		ct.activeHooks[hook], sp = ct.tr.Start(ctx, spanName, trace.WithAttributes(attrs...), trace.WithSpanKind(trace.SpanKindClient))
+		if ct.root == nil {
+			ct.root = sp
+		}
```

Comment:
If the end() is called before Start is finished then with the new logic span for this hook will never start and hence will not be recorded.

Reply:
The use of `ct.root` is a bit weird.  It's used as a place to attach HTTP headers and request error status in the `wroteHeaderField` and `wroteRequest` callbacks, but doesn't exist as a defined span of its own.  Instead, the first started span is attached to the `ct.root` member.  In both this logic and the prior logic nothing is assigned to `ct.root` until a span is started, so I don't think it's the case that the "root" span will never start, but it is possible that no span will have started before a request error is recorded.

I considered starting a separate span for the `ct.root` in `NewClientTrace()`, but it wasn't clear to me 1) whether that was an appropriate location as a client tracer may be constructed well before a request starts and 2) how to ensure that the span is ended.  Another alternative might be to use the span from the context provided to `NewClientTrace()` as `ct.root`, relying on the caller to end the span.

Reply:
@Aneurysm9 , I put the comment in wrong place. It was mean't for the `if` block. If the ct.end() finishes before ct.start() then the `if` block will not execute. Hence the span will not start/end.

Reply:
This was done this way initially in #283 to remove a panic from `ct.end()` in the case that a span hadn't already started.  Previously it stored a `NoopSpan` in the `activeHooks` map during the `else` block in `ct.end()`.  Since I changed the `activeHooks` map to store contexts instead of spans I changed the `else` block in `ct.end()` to store `context.Background()` which results in the span returned from `trace.SpanFromContext(hookCtx)` in `ct.start()` being a `NoopSpan` that is ended, but doesn't really need to be.

I think it was the expectation that those spans would not be recorded and there would not have been any discernable timeframe over which to record them.  I suppose we could alternately start and immediately end a span in `ct.end()` in this case, leaving a placeholder context to be cleaned up in `ct.start()` as a way of preventing it from starting a span that would never be ended.

Reply:
IMO, it is better to record the span even though there isn't a discernible timeframe. Not seeing the associated span for the trace would imply that ct.start() never happened but it did.

Reply:
Sure thing.  How does this look? Or would you prefer the span be ended in `ct.start()`, effectively inverting the recording?

Reply:
I think there is a race condition introduce if we record in `end` that could be avoided by adding locking around the `activeHooks` map, or like you are saying move the span recording to the `start`. +1 on moving to `start`.

Reply:
> Sure thing. How does this look? Or would you prefer the span be ended in `ct.start()`, effectively inverting the recording?

Works. The only thing is attributes may be missing in the end().

Reply:
> Works. The only thing is attributes may be missing in the end().

Yup, fixed that with https://github.com/open-telemetry/opentelemetry-go/pull/618/commits/2eac96e5fe4e56824f89d6a4f26d6676595ef587#diff-ad5e6fa2f2080149c8a5aebd94ea03ddR106